util/expvarx/expvarx.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package expvarx provides some extensions to the expvar package.
+package expvarx
+
+import (
+	"encoding/json"
+	"expvar"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"tailscale.com/types/lazy"
+)
+
+// SafeFunc is a wrapper around expvar.Func that guards against unbounded call
+// time and ensures that only a single call is in progress at any given time.
+type SafeFunc struct {
+	f      expvar.Func
+	limit  time.Duration
+	onSlow func(time.Duration, any)
+
+	mu       sync.Mutex
+	inflight *lazy.SyncValue[any]
+}
+
+// NewSafeFunc returns a new SafeFunc that wraps f. If f takes longer than limit
+// to execute then Value calls return nil. If non-nil, the parameter onSlow is a
+// function that is called when a slow operation completes, and is passed the
+// duration that f took to complete and the final value it produced.
+func NewSafeFunc(f expvar.Func, limit time.Duration, onSlow func(time.Duration, any)) *SafeFunc {
+	return &SafeFunc{f: f, limit: limit, onSlow: onSlow}
+}
+
+// Value acts similarly to expvar.Func.Value, but if the underlying function
+// takes longer than the configured limit, all callers will receive nil until
+// the underlying operation completes. On completion of the underlying
+// operation, the onSlow callback is called if set.
+func (s *SafeFunc) Value() any {
+	s.mu.Lock()
+
+	if s.inflight == nil {
+		s.inflight = new(lazy.SyncValue[any])
+	}
+	var inflight = s.inflight
+	s.mu.Unlock()
+
+	// inflight ensures that only a single work routine is spawned at any given
+	// time, but if the routine takes too long inflight is populated with a nil
+	// result. The long running computed value is lost forever.
+	return inflight.Get(func() any {
+		start := time.Now()
+		var wasSlow atomic.Bool
+		r := make(chan any, 1)
+
+		// work is spawned in routine so that the caller can timeout.
+		go func() {
+			// Allow new work to be started after this work completes
+			defer func() {
+				s.mu.Lock()
+				s.inflight = nil
+				s.mu.Unlock()
+			}()
+			v := s.f.Value()
+			r <- v
+			// If a caller believes we were slow, let onSlow know.
+			if wasSlow.Load() && s.onSlow != nil {
```

Comment:
How precise does this need to be?

We have a minor race.
* `r <- v` gets called, buffering the result on the channel
* in the select below, we hit the `<-time.After(s.limit)` case but stall.
* this goroutine read `wasSlow.Load()` and it reports false, so we never call `onSlow`.
* the other goroutine finally sets `wasSlow.Store(true)` and returns nil, but by then no one is listening.

Reply:
I'll also note: do we need a callback? could we just increment a global atomic metric?

Reply:
Both true. I had considered just a metric but as soon as we discover that one seems to be broken proving the theory, the next step is wanting to know "who dun it", and what they did. I moved onSlow to dispatch into a goroutine that drains the result channel that is conditionally started by the timeout path.